packages/pigeon/lib/cpp_generator.dart

```diff
@@ -0,0 +1,683 @@
+      indent.writeln('/** The codec used by ${api.name}. */');
+      indent.writeln('static const flutter::StandardMessageCodec& GetCodec();');
+      indent.writeln(
+          '/** Sets up an instance of `${api.name}` to handle messages through the `binaryMessenger`. */');
+      indent.writeln(
+          'static void Setup(flutter::BinaryMessenger* binaryMessenger, ${api.name}* api);');
+      indent.writeln(
+          'static flutter::EncodableMap WrapError(std::exception exception);');
+    });
+  });
+}
+
+void _writeHostApiSource(Indent indent, Api api) {
+  assert(api.location == ApiLocation.host);
+
+  final String codecName = _getCodecName(api);
+  indent.format('''
+/** The codec used by ${api.name}. */
+const flutter::StandardMessageCodec& ${api.name}::GetCodec() {
+\treturn flutter::StandardMessageCodec::GetInstance(&$codecName::GetInstance());
+}
+''');
+  indent.writeln(
+      '/** Sets up an instance of `${api.name}` to handle messages through the `binaryMessenger`. */');
+  indent.write(
+      'void ${api.name}::Setup(flutter::BinaryMessenger* binaryMessenger, ${api.name}* api) ');
+  indent.scoped('{', '}', () {
+    for (final Method method in api.methods) {
+      final String channelName = makeChannelName(api, method);
+      indent.write('');
+      indent.scoped('{', '}', () {
+        indent.writeln(
+            'auto channel = std::make_unique<flutter::BasicMessageChannel<flutter::EncodableValue>>(');
+        indent.inc();
+        indent.inc();
+        indent.writeln('binaryMessenger, "$channelName", &GetCodec());');
+        indent.dec();
+        indent.dec();
+        indent.write('if (api != nullptr) ');
+        indent.scoped('{', '} else {', () {
+          indent.write(
+              'channel->SetMessageHandler([api](const auto& message, auto reply)');
+          indent.scoped('{', '});', () {
+            final String returnType = _cppTypeForDartType(method.returnType);
+            indent.writeln('auto wrapped = flutter::EncodableMap();');
+            indent.write('try ');
+            indent.scoped('{', '}', () {
+              final List<String> methodArgument = <String>[];
+              if (method.arguments.isNotEmpty) {
+                indent.writeln(
+                    'auto args = std::get<flutter::EncodableList>(message);');
+                enumerate(method.arguments, (int index, NamedType arg) {
+                  final String argType = _cppTypeForDartType(arg.type);
+                  final String argName = _getSafeArgumentName(index, arg);
+                  indent.writeln(
+                      '$argType $argName = std::any_cast<$argType>(std::get<flutter::CustomEncodableValue>(args.at($index)));');
+                  /*indent.write('if ($argName == nullptr) ');
+                  indent.scoped('{', '}', () {
+                    indent.writeln(
+                        'throw exception("$argName unexpectedly null.");');
+                  });
+                  */
```

Comment:
Why is there commented-out code?

Reply:
This is for nullables, which is not implemented here yet. I've tried with optional<T>, but c++17 doesn't support optional references, so I think I'll go the simpler pointer route. This will then be uncommented.